File: core/pydataease-backend/app/models/data_filing.py
from __future__ import annotations

import time
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class FilingConfig(Base):
    """Data filing form configuration."""

    __tablename__ = "core_filing_config"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, default=lambda: time.time_ns())
    name: Mapped[str] = mapped_column(String(200), nullable=False, comment="Filing form name")
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="draft", comment="draft, published, disabled")
    target_datasource_id: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="Target datasource for writes")
    target_table: Mapped[str | None] = mapped_column(String(200), nullable=True, comment="Target table name")
    form_schema: Mapped[dict | None] = mapped_column(JSONB, nullable=True, default=dict, comment="Form field definitions and validation rules")
    field_mapping: Mapped[dict | None] = mapped_column(JSONB, nullable=True, default=dict, comment="Mapping from form fields to target table columns")
    idempotency_window_seconds: Mapped[int] = mapped_column(BigInteger, nullable=False, default=300, comment="Idempotency window in seconds")
    oid: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="Org scope")
    creator_uid: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="Creator user ID")
    create_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)
    update_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)


class FilingSubmission(Base):
    """Data filing submission record."""

    __tablename__ = "core_filing_submission"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, default=lambda: time.time_ns())
    filing_id: Mapped[int] = mapped_column(BigInteger, nullable=False, index=True, comment="Reference to FilingConfig")
    payload_hash: Mapped[str | None] = mapped_column(String(64), nullable=True, comment="SHA-256 hash of payload for idempotency")
    payload: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="Submitted data")
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending", comment="pending, success, failed, retrying")
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True, comment="Error details if failed")
    submitter_uid: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="Submitting user ID")
    retry_count: Mapped[int] = mapped_column(BigInteger, nullable=False, default=0)
    create_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)
    update_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)


class FilingAudit(Base):
    """Audit trail for all filing operations."""

    __tablename__ = "core_filing_audit"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, default=lambda: time.time_ns())
    filing_id: Mapped[int] = mapped_column(BigInteger, nullable=False, index=True, comment="Reference to FilingConfig")
    submission_id: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="Reference to FilingSubmission")
    action: Mapped[str] = mapped_column(String(30), nullable=False, comment="submit, publish, disable, retry, config_update")
    actor_uid: Mapped[int | None] = mapped_column(BigInteger, nullable=True, comment="User who performed the action")
    details: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="Action details")
    outcome: Mapped[str] = mapped_column(String(20), nullable=False, default="success", comment="success, failure")
    error_code: Mapped[str | None] = mapped_column(String(50), nullable=True, comment="Error code if failed")
    create_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)

File: core/core-frontend/src/api/data-filing.ts
import request from '@/config/axios'

export type FilingConfigStatus = 'draft' | 'published' | 'disabled'
export type SubmissionStatus = 'pending' | 'success' | 'failed' | 'retrying'
export type AuditAction = 'submit' | 'publish' | 'disable' | 'retry' | 'config_update'
export type AuditOutcome = 'success' | 'failure'

export interface FilingConfig {
  id: number
  name: string
  status: FilingConfigStatus
  targetDatasourceId: number | null
  targetTable: string | null
  formSchema: Record<string, any>
  fieldMapping: Record<string, any>
  idempotencyWindowSeconds: number
  oid: number | null
  creatorUid: number | null
  createTime: string | null
  updateTime: string | null
}

export interface FilingConfigCreateRequest {
  name: string
  targetDatasourceId?: number | null
  targetTable?: string | null
  formSchema: Record<string, any>
  fieldMapping: Record<string, any>
  idempotencyWindowSeconds?: number
  oid?: number | null
}

export interface FilingConfigUpdateRequest {
  name?: string | null
  targetDatasourceId?: number | null
  targetTable?: string | null
  formSchema?: Record<string, any> | null
  fieldMapping?: Record<string, any> | null
  idempotencyWindowSeconds?: number | null
}

export interface FilingSubmission {
  id: number
  filingId: number
  payloadHash: string | null
  payload: Record<string, any> | null
  status: SubmissionStatus
  errorMessage: string | null
  submitterUid: number | null
  retryCount: number
  createTime: string | null
  updateTime: string | null
}

export interface FilingAudit {
  id: number
  filingId: number
  submissionId: number | null
  action: AuditAction
  actorUid: number | null
  details: Record<string, any> | null
  outcome: AuditOutcome
  errorCode: string | null
  createTime: string | null
}

// Config CRUD
export const filingConfigList = (status?: string) =>
  request.get({ url: '/data-filing/config/list', params: { status } })

export const filingConfigGet = (id: number) => request.get({ url: `/data-filing/config/${id}` })

export const filingConfigCreate = (data: FilingConfigCreateRequest) =>
  request.post({ url: '/data-filing/config/create', data })

export const filingConfigUpdate = (id: number, data: FilingConfigUpdateRequest) =>
  request.put({ url: `/data-filing/config/${id}`, data })

export const filingConfigPublish = (id: number) =>
  request.post({ url: `/data-filing/config/${id}/publish` })

export const filingConfigDisable = (id: number) =>
  request.post({ url: `/data-filing/config/${id}/disable` })

export const filingConfigDelete = (id: number) =>
  request.delete({ url: `/data-filing/config/${id}` })

// Submissions
export const filingSubmit = (filingId: number, data: Record<string, any>) =>
  request.post({ url: `/data-filing/${filingId}/submit`, data })

export const filingSubmissions = (filingId: number) =>
  request.get({ url: `/data-filing/${filingId}/submissions` })

export const filingSubmissionGet = (submissionId: number) =>
  request.get({ url: `/data-filing/submission/${submissionId}` })

export const filingSubmissionRetry = (submissionId: number) =>
  request.post({ url: `/data-filing/submission/${submissionId}/retry` })

// Audit
export const filingAuditList = (filingId: number) =>
  request.get({ url: `/data-filing/${filingId}/audit` })
